src/edu/washington/escience/myria/operator/network/Producer.java

```diff
@@ -242,50 +251,57 @@ public void triggered(final IPCEvent event) {
     ioChannelsAvail[i] = true;
   }
 
-  /**
-   * set backup buffers as KeepMinValue.
-   * 
-   * @param keyColIndices the same as the one in KeepMinValue
-   * @param valueColIndices the same as the one in KeepMinValue
-   */
-  public void setBackupBufferAsMin(final int[] keyColIndices, final int[] valueColIndices) {
-    triedToSendTuples = new ArrayList<StreamingState>();
-    for (int i = 0; i < outputIDs.length; i++) {
-      triedToSendTuples.add(i, new KeepMinValue(keyColIndices, valueColIndices));
-      triedToSendTuples.get(i).setAttachedOperator(this);
+  @Override
+  public Schema getInputSchema() {
+    if (getChild() != null) {
+      return getChild().getSchema();
     }
+    return null;
   }
 
-  /**
-   * set backup buffers as KeepAndSortOnMinValue.
-   * 
-   * @param keyColIndices the same as the one in KeepAndSortOnMinValue
-   * @param valueColindices the same as the one in KeepAndSortOnMinValue
-   */
-  public void setBackupBufferAsPrioritizedMin(final int[] keyColIndices, final int[] valueColIndices) {
-    triedToSendTuples = new ArrayList<StreamingState>();
-    for (int i = 0; i < outputIDs.length; i++) {
-      triedToSendTuples.add(i, new KeepAndSortOnMinValue(keyColIndices, valueColIndices));
-      triedToSendTuples.get(i).setAttachedOperator(this);
-    }
+  @Override
+  public List<StreamingState> getStreamingStates() {
+    return triedToSendTuples;
   }
 
-  /** set backup buffers as DupElim. */
-  public void setBackupBufferAsDupElim() {
-    triedToSendTuples = new ArrayList<StreamingState>();
-    for (int i = 0; i < outputIDs.length; i++) {
-      triedToSendTuples.add(i, new DupElim());
-      triedToSendTuples.get(i).setAttachedOperator(this);
+  @Override
+  public void setStreamingStates(final List<StreamingState> states) {
+    triedToSendTuples = states;
+    for (StreamingState state : states) {
+      state.setAttachedOperator(this);
     }
   }
 
-  /** set backup buffers as SimpleAppender. */
-  public void setBackupBufferAsAppender() {
-    triedToSendTuples = new ArrayList<StreamingState>();
+  /**
+   * set backup buffers.
+   * 
+   * @param stateEncoding the streaming state encoding.
+   * 
+   * */
+  public void setBackupBuffer(final StreamingStateEncoding<?> stateEncoding) {
```

Comment:
Is it possible to extend `StreamingStateEncoding` with an abstract `getStreamingState()` method that delegates constructing the appropriate `StreamingState` to the corresponding `StreamingStateEncoding` subclass?